### ISSUE TITLE
Prototype fine-tuning the model on a new bone

### ISSUE THREAD

**Issue by richard-lane** [JGIBristol/Zebrafish_Osteoarthritis#1]
Fine tuning
====
Now that we have model(s) for segmenting out the jaw, we want to generalise this to work with other bones.

## Background

We will still need labelled training data for this, but the jaw model can serve as a starting point.
Instead of training a new model on the training data from scratch, we case base the new model on the old one.
Since it has already learned to segment out objects from a CT scan based on their location/shape/textures/etc.,
it will likely only need a relatively small amount of extra information to apply this knowledge to new bones.
This is transfer learning, or fine-tuning the model.

We have two models, so both of them will need to be fine-tuned:
 - the location model: locates the jaw in the CT scan
 - the segmentation model: segments the jaw out from the cropped scan

If your new bone data is already cropped, you won't need to fine-tune the location model.

## Things to do first
See `docs/` for help.
 - Make sure you can train the existing models (using the scripts in the `scripts/` directory).
 - Make sure you can run inference with these models
 - Find some new training data for transfer learning

## Steps
You can either write everything again from scratch, or use some of the existing code that I've left in this repo.
The advantage of writing everything from scratch is that you won't have to spend time trying to understand my existing code,
but you will probably end up re-implementing lots of things that are already there.
In either case you should look at the existing code for inspiration on things like the training loop/dataloader/etc.

I made a rough attempt at doing some of these scripts [here](https://github.com/JGIBristol/Zebrafish_Osteoarthritis/blob/main/scripts/experiments/quadrate_transfer_learning.py).

 - Make a new script
 - Optional: cache the training data as DICOMs. This will speed up development...
 - Fine-tune the location model
 - Fine-tune the segmentation model

You will have to work out the lines of code to write to do the above steps. The rough script linked above is an example of training
the segmentation model.

### Fine-tuning the models
For both models, the basic procedure is the same:
 - load the old (jaw) model into memory
 - make new training data - input and output (i.e. input image and object location, or input image + segmentation mask)
 - train the model - possibly with a lower learning rate than was used for the original models' training

I think the best way to figure out how to fine-tune the models will be to figure out how they were trained in the first place.
The original models are slightly unusual - read the `docs/` and look through the source code to find out:
 - how the training loop works (this is quite simple)
 - what the training data is (this is simple for the segmentation model, a bit complicated for the jaw location model)
 - if there's anything else going on (e.g. downsampling in the location model)

All their settings (hyperparameters etc.) are in `userconf.yml`.

Time Estimate, FTE / days
----
 - 5: familiarise with the project, read source code, get access to training data, read background papers, have meetings
 - 2 run the scripts: create DICOMs, train your own models, run inference with them
 - 5: create dataloader, fine-tune jaw location model.
 - 3: create dataloader, fine-tune segmentation model.

Total: 15 days

## Preliminary Plots
We don't have many quadrates here, I think 11 in the train set, so the result isn't great (could possibly be better if the LR was better tuned)

### Fine-tuning
Loss:

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/fe993e1a-8bd7-4732-906f-5332b59dd0b6" />

segmentated quadrate:

<img width="1427" height="475" alt="Image" src="https://github.com/user-attachments/assets/b7bd427b-fe77-4b13-98db-e1f45726ce6d" />

### Freshly trained
Loss:

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/00c4aa1f-8f42-4af9-816f-72f52fa29994" />

segmented quadrate (it didn't work well):

<img width="1427" height="475" alt="Image" src="https://github.com/user-attachments/assets/26140e91-31c7-4282-b3c0-1b61cfa5cc16" />

<img width="812" height="825" alt="Image" src="https://github.com/user-attachments/assets/961b453b-b227-47cd-80ec-b3345f6209af" />
